12-offer-encoding.md

```diff
@@ -0,0 +1,823 @@
+         \   /                                    |
+          v v                                     v
+  L1A2A=H("LnBranch",L1A||L2A)                   L3A=H("LnBranch",L3nonce||L3)
+  
+Assume L1A2A > L3A:
+
+  L1A2A=H("LnBranch",L1A||L2A)          L3A
+                          \            /
+                           v          v
+                Root=H("LnBranch",L3A||L1A2A)
+
+Signature = SIG("lightningoffersignature", Root, nodekey)
+```
+
+# Offers
+
+Offers are a precursor to an invoice: readers will either request an invoice
+(or multiple) or send an invoice based on the offer.  An offer can be much longer-lived than a
+particular invoice, so has some different characteristics; in particular it
+can be recurring, and the amount can be in a non-lightning currency.  It's
+also designed for compactness, to easily fit inside a QR code.
+
+The human-readable prefix for offers is `lno`.
+
+## TLV Fields for Offers
+
+1. `tlv_stream`: `offer`
+2. types:
+    1. type: 2 (`chains`)
+    2. data:
+        * [`...*chain_hash`:`chains`]
+    1. type: 6 (`currency`)
+    2. data:
+        * [`...*utf8`:`iso4217`]
+    1. type: 8 (`amount`)
+    2. data:
+        * [`tu64`:`amount`]
+    1. type: 10 (`description`)
+    2. data:
+        * [`...*utf8`:`description`]
+    1. type: 12 (`features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 14 (`absolute_expiry`)
+    2. data:
+        * [`tu64`:`seconds_from_epoch`]
+    1. type: 16 (`paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
+    1. type: 20 (`issuer`)
+    2. data:
+        * [`...*utf8`:`issuer`]
+    1. type: 22 (`quantity_min`)
+    2. data:
+        * [`tu64`:`min`]
+    1. type: 24 (`quantity_max`)
+    2. data:
+        * [`tu64`:`max`]
+    1. type: 30 (`node_id`)
+    2. data:
+        * [`point32`:`node_id`]
+    1. type: 54 (`send_invoice`)
+    1. type: 34 (`refund_for`)
+    2. data:
+        * [`sha256`:`refunded_payment_hash`]
+    1. type: 240 (`signature`)
+    2. data:
+        * [`bip340sig`:`sig`]
+
+1. subtype: `blinded_path`
+2. data:
+   * [`point`:`blinding`]
+   * [`byte`:`num_hops`]
+   * [`num_hops*onionmsg_path`:`path`]
+
+## Requirements For Offers
+
+A writer of an offer:
+  - if it sets `node_id`:
+    - MUST set `node_id` to the public key of the node to request the invoice from.
+  - otherwise:
+    - MUST provide at least one `blinded_path`
+    - MUST use the final `onionmsg_path` `point` in the first `blinded_path` as the implied `node_id` for `signature`.
+  - MAY specify exactly one signature TLV: `signature`:
+    - If so, it MUST set `sig` to the signature using `node_id` as described in [Signature Calculation](#signature-calculation).
+  - MUST set `description` to a complete description of the purpose
+    of the payment.
+  - if the chain for the invoice is not solely bitcoin:
+    - MUST specify `chains` the offer is valid for.
+  - otherwise:
+    - MAY omit `chains`, implying that bitcoin is only chain.
```

Comment:
Couple questions:

(1) Is this also meant to differentiate main net from testnet, etc?
(2) Is there a need to support multiple chains outside of bitcoin? Seems making this TLV a single `chain_hash` would simplify implementations.

Reply:
(1) yes
(2) If your implementation only supports bitcoin (which seems to be the case for all current implementations), you would just ignore all other chain hashes and it doesn't add any complexity to your implementation.

Reply:
> (2) If your implementation only supports bitcoin (which seems to be the case for all current implementations),

Is a TLV that includes both mainnet and testnet bitcoin valid?

> you would just ignore all other chain hashes and it doesn't add any complexity to your implementation.

The complexity comes from converting from a TLV stream to an abstract `Offer` type (and back). A simple implementation in rust would use an `Option<BlockHash>` field where the accessor would give the underlying `BlockHash` or the mainnet block hash if `None`. But if more than one chain is supported, an implementation would need to:

(1) determine if the offer contains a supported chain
(2) determine if there are any conflicts (mainnet and testnet bitcoin, for instance)
(3) convert `amount` to a supported chain if the first chain is unsupported
(4) store all chains in order to convert back to the same TLV stream

A simple implementation would only need to do (1). Arguably it would still do (4) if there is only one chain, but for multiple chains it makes checks for (1-3) more complex and makes it possible to represent an invalid state (2).

Reply:
> Is a TLV that includes both mainnet and testnet bitcoin valid?

It doesn't make much sense to do it but yes, it is valid.

> (1) determine if the offer contains a supported chain
> (2) determine if there are any conflicts (mainnet and testnet bitcoin, for instance)
> (3) convert `amount` to a supported chain if the first chain is unsupported
> (4) store all chains in order to convert back to the same TLV stream

I think you're overestimating the complexity:
(1) Just write `chains.contains(mySupportedChain)` instead of `chain == mySupportedChain`.
(2) There are no conflicts.
(3) If your wallet support multiple chains, you already need a mechanism to convert amounts between currencies. There is no new complexity here.
(4) Storing a few more bytes shouldn't be a big deal.

I'm not saying that allowing multiple chains is a must-have feature, I expect that most people (including myself) will only support a single bitcoin chain (which can be configured at compile time to be mainnet or testnet). I'm just saying that I can imagine some people wanting to support other chains in the future and it doesn't add complexity for the single-supported-chain implementations.

Reply:
Sidechains and the like are the main reason for this, which don't have the currency conversion issues.

Let's make this explicit: that if `amount` is specified, then all chains must use the same currency, and that the reader can assume that.

Reply:
No, we're going to want this, for stuff like fedimint too.  It's zero hassle if you don't support it.  You can't leave it to invoice_request time, since you don't know what kind of invoice you're allowed to ask for.